### PR TITLE
add context cancellation checks on GetSeries

### PR DESCRIPTION
**Pull request by erlan-z**
**What this PR does**:
Add some context cancellation checks on GetSeries

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -565,6 +570,9 @@ func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
 
 		// SeriesSet may have some series backed up by chunks, and some not.
 		for set.Next() {
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
```
Comment (alanprot) [inline pkg/querier/querier.go:575]: Why we dont return `storage.ErrSeriesSet(err)` here instead of in the caller?

Comment (erlan-z) [reply pkg/querier/querier.go]: Yes, we can do that. Changed

File: pkg/querier/series/series_set.go [hunk 2]
```diff
@@ -167,15 +168,18 @@ func MatrixToSeriesSet(sortSeries bool, m model.Matrix) storage.SeriesSet {
 }
 
 // MetricsToSeriesSet creates a storage.SeriesSet from a []metric.Metric
-func MetricsToSeriesSet(sortSeries bool, ms []metric.Metric) storage.SeriesSet {
+func MetricsToSeriesSet(ctx context.Context, sortSeries bool, ms []metric.Metric) (storage.SeriesSet, error) {
```
Comment (alanprot) [inline pkg/querier/series/series_set.go:171]: Same here?

```
return storage.ErrSeriesSet(err)
```

File: pkg/querier/querier.go [hunk 3]
```diff
@@ -565,6 +565,9 @@ func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
 
 		// SeriesSet may have some series backed up by chunks, and some not.
 		for set.Next() {
+			if ctx.Err() != nil {
+				return storage.ErrSeriesSet(ctx.Err())
+			}
```
Comment (alanprot) [inline pkg/querier/querier.go:570]: I think we did something similar and thanos and the perf was ok, right @yeya24 ?

Comment (yeya24) [reply pkg/querier/querier.go:570]: We have something like this when expanding postings. No obvious impact.

File: pkg/querier/querier.go [hunk 4]
```diff
@@ -553,7 +553,7 @@ func (querier) Close() error {
 	return nil
 }
 
-func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
+func (q querier) mergeSeriesSets(ctx context.Context, sets []storage.SeriesSet) storage.SeriesSet {
```
Comment (yeya24) [inline pkg/querier/querier.go:556]: I have another pr that removes this function totally https://github.com/cortexproject/cortex/pull/5825 so might affect this change.

File: pkg/querier/series/series_set.go (hunk 2 above)
Comment (erlan-z) [reply pkg/querier/series/series_set.go]: Changed

**Review by alanprot — APPROVED** [review]

File: pkg/querier/querier.go (hunk 4 above)
Comment (erlan-z) [reply pkg/querier/querier.go]: removed the change. It will be removed with above PR anyway.

**Review by yeya24 — APPROVED** [review]
Thanks!

**Comment by yeya24** [comment]
@erlan-z Can you update changelog please?

**Comment by yeya24** [comment]
I feel we might also miss some context check at `storage.NewMergeSeriesSet`. But let's see if this problem still reproducible after this change and we can improve later.

**Pull request merged**